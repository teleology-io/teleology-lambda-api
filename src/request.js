import qs from 'querystring';
import { clean } from '@teleology/fp';

const extractBody = ({ body, isBase64Encoded }) => {
  let value = body;
  if (isBase64Encoded) {
    try {
      value = Buffer.from(body, 'base64');
    } catch (e) {
      // do nothing
    }
  }

  try {
    const o = JSON.parse(value);
    if (o && typeof o === 'object') {
      return o;
    }
  } catch (e) {
    // do nothing
  }

  try {
    return { ...qs.parse(value) };
  } catch (e) {
    // do nothing
  }

  return body;
};

export default ({
  headers = {},
  pathParameters = {},
  queryStringParameters = {},
  body = '',
  isBase64Encoded,
  requestContext,
  ...event
}) => {
  const h = Object.keys(headers).reduce((a, key) => {
    a[key.toLowerCase()] = headers[key];
    return a;
  }, {});

  const ctx = {};

  if (requestContext) {
      // Claims portion here is to support serverless-offline  
      ctx.auth = requestContext?.authorizer?.claims || requestContext.authorizer;
  }

  return clean({
    ...event,
    ...ctx,
    headers: h,
    requestContext,
    data: {
      ...pathParameters,
      ...queryStringParameters,
      ...extractBody({ body, isBase64Encoded }),
    },
  });
};
