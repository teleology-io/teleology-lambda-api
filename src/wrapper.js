/* eslint-disable no-param-reassign */
import request from './request';
import response from './response';

const serializeError = (e) =>
  Object.getOwnPropertyNames(e).reduce((error, key) => {
    // Don't include stack in json response
    if (key === 'stack') return error;

    error[key] = e[key];
    return error;
  }, {});

export default (fn) => async (event) => {
  try {
    const resolved = await fn(request(event));
    if (Array.isArray(resolved)) {
      return response({ body: resolved });
    }

    const { code = 200, body, headers = {}, ...rest } = resolved || {};
    return response({
      code,
      headers,
      body: body || rest,
    });
  } catch (e) {
    return response({
      code: e.code || 500,
      body: serializeError(e),
    });
  }
};
