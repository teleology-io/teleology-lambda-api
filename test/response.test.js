import { response } from '../src';

describe('response', () => {
  it('uses defaults', () => {
    const result = response({});

    expect(result).toEqual({
      body: '',
      headers: {
        'Access-Control-Allow-Credentials': true,
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      },
      statusCode: '200',
    });
  });

  it('stringifies body data', () => {
    const result = response({
      body: {
        foo: 'bar',
      },
    });

    expect(result.body).toEqual(JSON.stringify({ foo: 'bar' }));
  });

  it('overrides headers', () => {
    const result = response({ headers: { 'Content-Type': 'text/plain' } });

    expect(result.headers['Content-Type']).toEqual('text/plain');
  });

  it('overrides code', () => {
    const result = response({ code: 500 });

    expect(result.statusCode).toEqual('500');
  });

  it('responds with array', () => {
    const result = response({ code: 200, body: [{ foo: 'bar' }] });

    expect(result.statusCode).toEqual('200');
    expect(result.body).toMatchSnapshot();
  });
});
