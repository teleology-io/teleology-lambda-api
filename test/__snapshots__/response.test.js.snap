// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`response responds with array 1`] = `"[{\\"foo\\":\\"bar\\"}]"`;
